GG/src/TextControl.cpp

```diff
@@ -204,6 +204,40 @@ void TextControl::SetText(const std::string& str)
     m_cached_minusable_size_width = X0;
 }
 
+void TextControl::SetText(const std::string& str,
+                          const std::vector<boost::shared_ptr<Font::TextElement> >& text_elements)
+{
+    if (!utf8::is_valid(str.begin(), str.end()))
+        return;
+    m_text = str;
+
+    m_code_points = CPSize(utf8::distance(str.begin(), str.end()));
+
+    m_text_elements = text_elements;
+    for (std::vector<boost::shared_ptr<Font::TextElement> >::iterator it = m_text_elements.begin();
+         it != m_text_elements.end(); ++it)
+    {
+        (*it)->Bind(m_text);
+    }
+
+    if (!m_font)
+        return;
+
+    m_line_data = m_font->DetermineLines(m_text, m_format, ClientSize().x, m_text_elements);
```

Comment:
Two things.  

1. PR #989 has a commit [Refactor TextControl::RecomputeLineData from SetText and ChangeTemplatedText](https://github.com/freeorion/freeorion/pull/989/commits/625341fd681ceebd74323f98f12573a037405239) , which refactors the repeated code into a function called `RecomputeLineData`.

2. This `DetermineLines()` function has had the slow parsing  code moved into `ExpensiveParseFromTextToTextElements()`.  It is only computing the `m_line_data, which means calculating where the line breaks are and adding the justification information to the line data so that prerender can place the glyphs in the line later.